Assign interactionID to simulated clicks

francovs opened the pull request:
#### e6a4c101ac293ba1cadfdfe2162dbd92a4465eea
<pre>
Assign interactionID to simulated clicks
<a href="https://rdar.apple.com/159873276">rdar://159873276</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=298409">https://bugs.webkit.org/show_bug.cgi?id=298409</a>

Reviewed by Ryosuke Niwa.

When computing interactionID for click events, we now consider whether
the click is being simulated by a keypress. This behavior is required
for some WPT tests and is consistent with Chrome and Firefox, but is
currently not in accordance with the spec. An issue on the spec has
been filed about this: <a href="https://github.com/w3c/event-timing/issues/147.">https://github.com/w3c/event-timing/issues/147.</a>

To implement this, a data member containing the interaction ID was
added to UIEvent.

This patch also adds a workaround to a bug with IME keydowns on
finalizeEventTimingEntry(). See <a href="https://bugs.webkit.org/show_bug.cgi?id=165004">https://bugs.webkit.org/show_bug.cgi?id=165004</a>

Now passing:
* LayoutTests/imported/w3c/web-platform-tests/event-timing/interactionid-keyboard-event-simulated-click-button-space-expected.txt
* LayoutTests/imported/w3c/web-platform-tests/event-timing/interactionid-keyboard-event-simulated-click-checkbox-space-expected.txt
* LayoutTests/imported/w3c/web-platform-tests/event-timing/interactionid-keyboard-event-simulated-click-link-enter-expected.txt

Canonical link: <a href="https://commits.webkit.org/299973@main">https://commits.webkit.org/299973@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7071f7697bafccd2ccc2254fd1def6758591d41d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120928 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40622 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31278 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127338 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73004 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122804 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41320 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49199 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91857 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61102 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123880 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32996 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108407 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72542 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32035 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26508 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70930 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102507 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26688 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130196 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47849 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36366 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100475 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48218 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104580 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100370 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25443 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45766 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23824 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/44540 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47711 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53416 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47182 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50526 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48866 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/97f9aad95b376936649bbac3a2a99d6a0a3808dc. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50354

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0c206078de372d7369b5f68d8a10c7df29dd889b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50354

francovs committed:
Assign interactionID to simulated clicks
rdar://159873276
https://bugs.webkit.org/show_bug.cgi?id=298409

Reviewed by Ryosuke Niwa.

When computing interactionID for click events, we now consider whether
the click is being simulated by a keypress. This behavior is required
for some WPT tests and is consistent with Chrome and Firefox, but is
currently not in accordance with the spec. An issue on the spec has
been filed about this: https://github.com/w3c/event-timing/issues/147.

To implement this, a data member containing the interaction ID was
added to UIEvent.

This patch also adds a workaround to a bug with IME keydowns on
finalizeEventTimingEntry(). See https://bugs.webkit.org/show_bug.cgi?id=165004

Now passing:
* LayoutTests/imported/w3c/web-platform-tests/event-timing/interactionid-keyboard-event-simulated-click-button-space-expected.txt
* LayoutTests/imported/w3c/web-platform-tests/event-timing/interactionid-keyboard-event-simulated-click-checkbox-space-expected.txt
* LayoutTests/imported/w3c/web-platform-tests/event-timing/interactionid-keyboard-event-simulated-click-link-enter-expected.txt

Canonical link: https://commits.webkit.org/299973@main
